src/main/scala/fi/oph/koski/ytl/YtlService.scala

```diff
@@ -0,0 +1,71 @@
+package fi.oph.koski.ytl
+
+import java.sql.Timestamp
+import java.time.LocalDateTime
+
+import fi.oph.koski.config.KoskiApplication
+import fi.oph.koski.db.{KoskiTables, OpiskeluoikeusRow}
+import fi.oph.koski.henkilo.OppijaHenkilö
+import fi.oph.koski.history.OpiskeluoikeusHistoryPatch
+import fi.oph.koski.http.HttpStatus
+import fi.oph.koski.json.{JsonSerializer, SensitiveDataAllowed}
+import fi.oph.koski.koskiuser.KoskiSpecificSession
+import fi.oph.koski.log._
+import fi.oph.koski.opiskeluoikeus.{OpiskeluoikeusQueryContext, OpiskeluoikeusQueryFilter, QueryOppijaHenkilö}
+import fi.oph.koski.organisaatio.MockOrganisaatiot.ytl
+import fi.oph.koski.schema.{Henkilö, KoskiSchema, Oppija}
+import fi.oph.scalaschema.extraction.ValidationError
+import org.json4s.JsonAST.JValue
+import org.json4s.MappingException
+import rx.lang.scala.Observable
+import fi.oph.scalaschema.{SchemaValidatingExtractor, SerializationContext, Serializer}
+
+class YtlService(application: KoskiApplication) extends Logging {
+  private val serializationContext = SerializationContext(KoskiSchema.schemaFactory, KoskiSchema.skipSyntheticProperties)
+
+  def streamOppijat(oidit: Seq[String], hetut: Seq[String], user: KoskiSpecificSession): Observable[JValue] = {
+    // Käytä globaaleja oikeuksia tietojen hakemiseen: filtteröinti YTL:lle sallittuihin kenttiin
+    // tapahtuu deserialisoinnissa ja muussa koodilogiikassa.
+    implicit val systemUser: KoskiSpecificSession = KoskiSpecificSession.systemUser
+
+    val henkilöt =
+      application.opintopolkuHenkilöFacade.findOppijatNoSlaveOids(oidit) ++
+        application.opintopolkuHenkilöFacade.findOppijatByHetusNoSlaveOids(hetut)
+    val oidToHenkilo: Map[Henkilö.Oid, OppijaHenkilö] = henkilöt.map(h => h.oid -> h).toMap
+    val masterOids = henkilöt.map(_.oid)
+    val queryFilters = List(OpiskeluoikeusQueryFilter.OppijaOidHaku(masterOids ++ application.henkilöCache.resolveLinkedOids(masterOids)))
+
+    OpiskeluoikeusQueryContext.streamingQueryGroupedByOid(application, queryFilters, None)
+      .map((t: (QueryOppijaHenkilö, List[OpiskeluoikeusRow])) => {
+        val hlö = oidToHenkilo(t._1.oid)
+        YtlOppija(
+          henkilö = YtlHenkilö(hlö, hlö.äidinkieli.flatMap(k => application.koodistoViitePalvelu.validate("kieli", k.toUpperCase))),
+          opiskeluoikeudet = t._2
+            .map(toYtlOpiskeluoikeus).flatMap(_.poistaTiedotJoihinEiKäyttöoikeutta)
+        )
+      })
+      // Poista listalta oppijat, joilla ei ollut yhtään (YTL:ää kiinnostavaa) opiskeluoikeutta
```

Comment:
Kuinkakohan paljon näitä YTL:ää kiinnostamattomia tulee olemaan? Ts. jos kuorma on ongelma, voiko siinä säästää paljonkin jos leikkaa tietyt opiskeluoikeustyypit pois? Tuleeko tässä esim kaikki peruskoulun opiskeluoikeudet nyt mukana?

Reply:
Hyvä huomio. Kannattaa varmasti välttää tietokantakyselystä alkaen muut kuin YTL:ää kiinnostavat opiskeluoikeustyypit.

Reply:
Tämä tehty.